src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+    binary sequence representing actual stored number.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.math.WideInt;
+
+import ocean.transition;
+import ocean.core.Test;
+import ocean.core.Enforce;
+
+/// ditto
+public struct WideInt ( size_t N )
+{
+    static assert (
+        N > 2,
+        "For 'N <= 2' using 'uint' or 'ulong' directly is suggested"
+    );
+
+    private uint[N] payload;
+
+    /***************************************************************************
+
+        Constructor from a regular ulong
+
+    ***************************************************************************/
+
+    this ( ulong value )
+    {
+        this.payload[0] = cast(uint) value;
+        this.payload[1] = value >> 32;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            amount of meaninfful decimal digits in currently stored value
+
+    ***************************************************************************/
+
+    public size_t digits ( )
+    {
+        WideInt copy = *this;
+        size_t count;
+        while (copy != 0)
+        {
+            copy.divideBy(10);
+            ++count;
+        }
+        return count;
+    }
+
+    /***************************************************************************
+
+        Inefficient allocating string conversion useful for tests and
+        prototyping.
+
+        Returns:
+            string representation
+
+    ***************************************************************************/
+
+    public istring toString ( )
+    {
+        mstring result;
+        this.toString((cstring s) { result ~= s; });
+        return assumeUnique(result);
+    }
+
+    unittest
+    {
+        WideInt num = 42;
+        test!("==")(num.toString(), "42");
+
+        num.payload[2] = 1;
+        test!("==")(num.toString(), "18446744073709551658");
+    }
+
+    /***************************************************************************
+
+        Sink based string conversion
+
+        Params:
+            sink = delegate to call with resulting string
+
+    ***************************************************************************/
+
+    public void toString ( void delegate (cstring) sink )
+    {
+        auto n = this.digits();
+        static mstring buffer;
+        buffer.length = n;
+        enableStomping(buffer);
+
+        WideInt copy = *this;
+        for (ptrdiff_t idx = n-1; idx >= 0; --idx)
+        {
+            auto remainder = copy.divideBy(10);
+            buffer[idx] = cast(char) ('0' + remainder);
+        }
```

Comment:
Sorry, but I can't help myself:
```d
foreach_reverse (ref c; buffer)
{
    c = ...
}
```
is just easier to read than
```d
for (ptrdiff_t idx = n-1; idx >= 0; --idx)
{
    buffer[idx] = ...
}
```
and inherently safe from off-by-one and index overflow errors.